app/models/web_of_science_source_record.rb

```diff
@@ -0,0 +1,46 @@
+class WebOfScienceSourceRecord < ActiveRecord::Base
+
+  validates :database, :source_data, :source_fingerprint, :uid, presence: true
+
+  after_initialize :extract_fields
+
+  # @return doc [Nokogiri::XML::Document] XML document
+  def doc
+    @doc ||= Nokogiri::XML(source_data)
+  end
+
+  # @return xml [String] XML
+  def to_xml
+    doc.to_xml(save_with: XML_OPTIONS).strip
+  end
+
+  private
+
+    XML_OPTIONS = Nokogiri::XML::Node::SaveOptions::AS_XML | Nokogiri::XML::Node::SaveOptions::NO_DECLARATION
+
+    def extract_fields
+      raise 'Missing source_data' if source_data.nil?
+      extract_active if is_active.nil?
+      extract_uid if uid.nil?
+      extract_database if database.nil?
+      extract_source_fingerprint if source_fingerprint.nil?
+    end
+
+    def extract_active
+      # assume records are active until we discover a deprecation attribute
+      self.is_active = true
+    end
+
+    def extract_database
+      uid_split = uid.split(':')
+      self.database = uid_split.length > 1 ? uid_split[0] : nil
+    end
+
+    def extract_source_fingerprint
+      self.source_fingerprint = Digest::SHA2.hexdigest(source_data)
+    end
+
+    def extract_uid
```

Comment:
I don't think we need separate methods for `extracting` attributes.  Just wrap the (public) ActiveRecord accessor methods, like:
```ruby
def uid
  super || self.uid = doc.search('UID').text
end
```
Then we should be able to rely on the rails validator as normal (in a lazy fashion) without the `after_initialize` callback.

Reply:
I'll try this, haven't seen this pattern before (been a long time since I've created an AR model). 
 If we don't use an `after_initialize`, I wonder if we need a `before_validate` or `before_save` callback to extract the field data?

Reply:
Trying this out, but not sure this will work with the `database` field (or any other `nil` valued field), because when it should be `nil` (or NULL), it will always keep resetting itself and possibly updating the record when it shouldn't, i.e. consider this when it should be `nil`:
```ruby
    def database
      super || begin
        uid_split = uid.split(':')
        self.database = uid_split.length > 1 ? uid_split[0] : nil
      end
    end
```

Reply:
rubocop complains about `found == in conditional`